pkg/webhook/service_labels/patching.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2020 Clastix Labs.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service_labels
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	capsulev1alpha1 "github.com/clastix/capsule/api/v1alpha1"
+	capsulewebhook "github.com/clastix/capsule/pkg/webhook"
+	"gomodules.xyz/jsonpatch/v2"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1beta1 "k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// +kubebuilder:webhook:path=/mutate-v1-service-labels,mutating=true,failurePolicy=ignore,groups="";discovery.k8s.io,resources=services;endpoints;endpointslices,verbs=create;update,versions=v1;v1beta1,name=service.labels.capsule.clastix.io
+
+type webhook struct {
+	handler capsulewebhook.Handler
+}
+
+func Webhook(handler capsulewebhook.Handler) capsulewebhook.Webhook {
+	return &webhook{handler: handler}
+}
+
+func (w webhook) GetHandler() capsulewebhook.Handler {
+	return w.handler
+}
+
+func (w webhook) GetName() string {
+	return "ServiceLabels"
+}
+
+func (w webhook) GetPath() string {
+	return "/mutate-v1-service-labels"
+}
+
+type handler struct {
+}
+
+func Handler() capsulewebhook.Handler {
+	return &handler{}
+}
+
+func (h *handler) OnCreate(clt client.Client, decoder *admission.Decoder) capsulewebhook.Func {
+	return func(ctx context.Context, req admission.Request) admission.Response {
+		return h.syncLabels(ctx, req, clt, decoder)
+	}
+}
+
+func (h *handler) OnUpdate(clt client.Client, decoder *admission.Decoder) capsulewebhook.Func {
+	return func(ctx context.Context, req admission.Request) admission.Response {
+		return h.syncLabels(ctx, req, clt, decoder)
+	}
+}
+
+func (h *handler) OnDelete(client client.Client, decoder *admission.Decoder) capsulewebhook.Func {
+	return func(ctx context.Context, req admission.Request) admission.Response {
+		return admission.Allowed("")
+	}
+}
+
+func (h *handler) syncLabels(ctx context.Context, req admission.Request, client client.Client, decoder *admission.Decoder) admission.Response {
+	var availableLabels, availableAnnotations map[string]string
+	var namespaceName string
+	var patch []jsonpatch.JsonPatchOperation
+
+	switch req.Kind.Kind {
+	case "Service":
+		svc := &corev1.Service{}
```

Comment:
I don't want to be picky but I guess the most elegant solution would be a type embedding in order to decorate these types and void all of these duplicate lines of codes [as I did here](https://github.com/clastix/capsule/blob/a4b0c0fe27906621af61093e05964f9a676f0c01/pkg/webhook/ingress/types.go#L17-L71)

Tl;dr; embedding Service, Endpoints and EndpointSlice into a domain struct adhering to an Interface providing some funcs as `GetNamespace`, `GetAnnotations`, `GetLabels`.

Reply:
thanks for the good advice, done :)